feat(BV): Interval domains for bit-vectors

bclement-ocp opened the pull request:
This patch adds interval domains to the Bitv_rel module, as well as
interreductions between the bitlist and interval domains following:

  Sharpening Constraint Programming approaches for Bit-Vector Theory.
  Zakaria Chihani, Bruno Marre, François Bobot, Sébastien Bardin.
  CPAIOR 2017. International Conference on AI and OR Techniques in
    Constraint Programming for Combinatorial Optimization Problems, Jun
    2017, Padova, Italy.

More precisely:

 - The `Intervals` module is extended to support the `extract` operation, which is used to propagate between bit-vector compositions and their components;

 - The interreductions are implemented using the new `Bitlist.increase_lower_bound`, `Bitlist.decrease_upper_bound`, and the new `shared_msb` helper in `Bitv_rel`;

 - Propagations are performed by alternating propagations until fixpoint in each domain, followed by interreductions and propagations until fixpoint in the other domain, until reaching a fixpoint for the whole procedure. It is not clear that this is the best strategy; the goal is to try and limit interreductions since they are relatively expensive but we should revisit this once more operations are supported.

For now, only the `bvule`, `bvult`, `bvugt` and `bvuge` primitives are supported as built-in bit-vector operations; other operations such as `bvadd` are still encoded using `bv2nat`. These operations will be migrated to bit-vector primitives in a follow-up PR.

Finally, there are some tests for the tricky bits (`Intervals.extract` and the interreduction primitives) using QCheck.

bclement-ocp commented:
Note: this is +107-515 or a net regression of about 400 problems on our QF_BV subset. This is expected: this PR uses bit-vector domains for `bvule` / `bvult` / etc. but does not change the use of `bv2nat` / `int2bv` round-trip for other arithmetic operators. This means that we are prevented from doing many interesting propagations.

A follow-up PR (should be ready sometime next week) fixes that by implementing arithmetic propagators for the BV interval domains: it is currently +1156-209 (or an improvement of almost 1000 problems) on that same subset compared to `next`, at least in part due to interreduction between arithmetic and bitwise constraints.

Halbaroth commented on src/lib/reasoners/intervals.ml:
```diff
@@ -1200,6 +1200,150 @@ let pick ~is_max { ints; is_int; _ } =
+    [f] *MUST* be monotone (either increasing or decreasing depending on the
+    value of the [increasing] labelled argument)
+
+    The [minfty] and [pinfty] optional arguments provide values for [map_bigint]
```
What's `map_bigint` here? I guess you meant `f` and `map_bigint` is a deprecated reference to `map_increasing_bigint` ?

bclement-ocp replied:
Yeah, that should be `f`, thanks (the name of the argument to this function used to be `map_bigint` instead of `f`).

Halbaroth commented on src/lib/reasoners/intervals.ml:
```diff
@@ -1200,6 +1200,150 @@ let pick ~is_max { ints; is_int; _ } =
+        (lb, ub) :: l
+      ) [] uints.ints
+  in
+  let res = union_intervals { uints with ints } in
+  assert (res.ints != []);
```
I don't understand why we need to normalize the result before returning it. If `f` is an increasing function, it sends an union of disjoint intervals to an union of disjoint intervals. If `f` is decreasing, we only need to reverse the list `ints`. I'm wrong?

Halbaroth replied:
If it's to be sure the returned domain is normalized, we may replace this by a heavy assert.

bclement-ocp replied:
We need to normalize the result because `f` is not guaranteed to be *strictly* increasing/decreasing.

So for instance maybe we have `[10, 12] U [14, 20]` but `f(x) = 10` for any `x ≥ 10` so after the call we get `[10, 10] U [10, 10]` which is not properly normalized.

We need to support this because `shr` is not strictly increasing (all sufficiently small values map to `0`).

(interestingly enough, I believe this is the case where `union_intervals` currently has a subtle soundness bug (: )

Halbaroth replied:
In English `increasing`means `strictement croissant`. So we should write [non-decreasing](https://fr.wiktionary.org/wiki/non-decreasing) :)

bclement-ocp replied:
> In English increasingmeans strictement croissant. So we should write [non-decreasing](https://fr.wiktionary.org/wiki/non-decreasing) :)

Unfortunately the usage varies and `increasing` is used for both. Monotone is the non-strict version and this is what the function is called. I will clarify in the comment

Halbaroth replied:
`strictly increasing` is acceptable ;)

bclement-ocp commented:
After reading and thinking a lot (and starting a small Coq developement) about the explanations for intervals:

 - I am quite confident that the `map_monotone` function introduced in this PR, and the `map2_monotone` function introduced in #1085, are correct
 - I am reasonably confident that the other interval functions introduced in this PR and following PRs are correct when called on a single interval, with the exception maybe of `extract`
 - I am *not* convinced that the other interval functions are correct when called on an union of intervals.

That is not to say that I don't think they are correct (in particular, I am confident that all of these functions return the right values for the bounds), but that they may forget necessary explanations in some cases, causing unsoundness. I think the cases where we forget explanations are quite rare (or may be made impossible currently) so I will rewrite these functions to try to be less clever and add more explanations. I will also write a documentation for the intervals module, now that i understand the invariants better.

This does not impact the rest of the PR or the rest of the PRs in the stack.

Halbaroth commented on tests/bitv/bitlist-interval001.smt2:
```diff
@@ -0,0 +1,5 @@
+(set-logic ALL)
```
Nit: the logic `QF_BV` is sufficient here?

bclement-ocp replied:
I tend to default to using `(set-logic ALL)` now because I copy-paste tests to test things during development or debugging and Dolmen complains if you use `QF_BV` but later use quantifiers or integers.

(Also Why3 always uses `(set-logic ALL)` so we might as well test with this)

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -77,6 +77,94 @@ let is_bv_ty = function
+
+  let map_leaves f r =
+    List.fold_left (fun ival { Bitv.bv; sz } ->
+        let ival = Intervals.scale (Q.of_bigint @@ Z.(~$1 lsl sz)) ival in
```
I'm not sure to understand this line. My understanding is as follows:
`map_leaves` builds an interval domain for the bitvector `r` by building interval domains of its leaves. For instance, says that `r` is the bitvector `[|10101|] + x + (extract i j k y)`. We build an interval domain for `[|10101|]`, `x` and `(extract i j k y)` and build the interval domain of `r` from them.

The function `Intervals.add i_1 i_2` computes the sum set `i_1 + i_2`, that is the direct image of `(i_1, i_2)` by `(x, y) |--> x + y`. 

The function `Intervals.scale s i` computes the direct image of the union of intervals `i` by the homothety  `x |--> s * x`.

We starts the construction with the interval singleton `0`, which makes sense as this set is the neutral element for the operator `Intervals.add`  on the union of intervals.

We have to make room for the interval domain of `{ Bitv.bv; sz }` in `ival` and I guess it's the purpose of the call to `Intervals.scale` here. It works if we fold on `r` from the most significant bit to the least one. Is it the case? I cannot remember if the bitvector are represented from the most significant bit to the least one in `bitv`?

bclement-ocp replied:
The `map_leaves` function operates on the structural representation of a bit-vector, so I guess `+` here means `concat`?

If in terms of concatenation we have `[| 10101 |] ++ x ++ (extract i j y)`, in terms of addition we have `[| 10101|] * 2^(sz x + j - i + 1) + x * 2^(j - i + 1) + (extract i j y)`. The scale is to bring the value in the appropriate range.

Bit-vectors are indeed stored most significant bit first (but indexed by least-significant bit first, which is always easy to mess up).

Halbaroth replied:
> Bit-vectors are indeed stored most significant bit first (but indexed by least-significant bit first, which is always easy to mess up).

I think we should mention this detail of implementation in the documentation of `Bitv`.

bclement-ocp replied:
It is mentioned in the documentation of `extract` already but we could move it to the documentation of `abstract` I suppose (not in this PR tho).

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -149,11 +237,17 @@ module Constraint : sig
       The explanation [ex] justifies that the constraint [t] applies, and must
       be added to any domain that gets updated during propagation. *)
+
+  val propagate_interval : ex:Ex.t -> t -> Interval_domains.Ephemeral.t -> unit
```
I know it's the same documentation of `propagate` but we should copy it or add a pointer to it.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+
+    touch_pending queue;
+    HX.iter (fun r () ->
+        HX.replace bitlist_changed r ();
```
Why not using a set of semantic values instead of a map?

bclement-ocp replied:
Because access to a set is `O(log n)` while a hash map is `O(1)`. There is no hash set in OCaml.

Halbaroth replied:
oh sorry, I was thinking about hash set but there is no such thing!

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -77,6 +77,94 @@ let is_bv_ty = function
+              ignore @@
+              intersect ~ex:Ex.empty int
+                (Intervals.point (Q.of_bigint z) Tint Ex.empty);
+              acc
```
I know it's not a part of this PR but I wonder if we could implement a better iterator:
1. When I see a fold iterator raising an exception, I expect that the exception is raised by the iterated function, not the iterator (or the iterator may raise a very specific exception as `List.fold_left2` for instance).

2. We only call `Domain.fold_leaves` to update leaves' domain of a semantic value `r` and to detect inconsistency during the structural propagation:
```ocaml
Domain.fold_leaves (fun r d () -> update r d) r (get r) ()
``` 
As we don't store bitlists for constant bitvectors, `fold_leaves` performs the update step (which consists in checking the consistency) when it sees a constant bitvector in `r`.

It means that `fold_leaves` isn't a generic iterator and we have to read the implementation of both `fold_leaves` and `f` to understand what's happened.

bclement-ocp replied:
Yeah, `fold_leaves` is poorly named, I agree. It is indeed not supposed to be a generic iterator; it should have a name that references (structural) propagation. I propose to change that in another PR though.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+    Bitlist.concat shared_bl (Bitlist.unknown (sz - nshared) Ex.empty)
+
+(* These are helpers for readability because [Intervals.mk_closed] makes my
+   brain melt *)
```
:D

bclement-ocp commented:
> I am quite confident that the map_monotone function introduced in this PR, and the map2_monotone function introduced inhttps://github.com/OCamlPro/alt-ergo/pull/1085, are correct

Turns out they are not (: I thought I had replicated the behavior of the `add` function properly, but turns out I missed a crucial detail.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+      ex ex Tint
+
+(* Algorithm 1 from "Sharpening Constraint Programming approaches for
+   Bit-Vector Theory". *)
```
We should add an example as it does in the article. I know we can read the article to find an example but it's more convenient to inline it here.

bclement-ocp replied:
Done

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+   CPAIOR 2017. International Conference on AI and OR Techniques in
+   Constraint Programming for Combinatorial Optimization Problems, Jun
+   2017, Padova, Italy.
+   https://cea.hal.science/cea-01795779/document *)
```
Same remark, we should add an example here.

bclement-ocp replied:
Done

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+         ignore the bounds' explanations. *)
+      let acc =
+        match lb with
+        | None -> assert false
```
I guess it cannot happen because our interval domains are always subsets of the interval `[0, n)` for some integer `n`.

bclement-ocp replied:
Yep. I guess I could treat this case as `Some (0, 0)`, but I'd rather be strict. I will add a comment explaining.

bclement-ocp replied:
This code has changed and now uses the new `finite_lower_bound` and `finite_upper_bound` helpers w/o assertions.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+      in
+      let acc =
+        match ub with
+        | None -> assert false
```
Same remark.

bclement-ocp replied:
Same as above

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -424,74 +642,284 @@ module Any_constraint = struct
+  List.fold_left (fun acc ((lb, _ex_lb), (ub, _ex_ub)) ->
+      (* Note: the bounds adjustment is expressed by removing specific
+         intervals that are forbidden by the bitlist domain. This is always
+         correct independently of the bounds' explanations: we only use the
```
This comment is a bit confusing. I understood that we drop these explanations in the new interval domain, which is clearly incorrect.

bclement-ocp replied:
What is happening is:

 - We have a bitlist domain that has an explanation [ex].
 - We build some intervals that the bitlist domain proves are impossible (for instance, if the bitlist domain is `00??`, then the interval `[4, 15]` is impossible because no value in the interval `[4, 15]` starts with `00`).
 - We are removing those intervals from the existing interval domain using `exclude`

The intervals that we build from the bitlist domain only need the explanation from the bitlist domain to be valid, so we don't need to incorporate the explanations from the old bounds into it. The `exclude` and `intersect` functions from the intervals module takes care of appropriately (hopefully… but if not it is a bug in `exclude`, not in this code) updating the bounds to take into accound both explanations.

Is that more clear?

bclement-ocp replied:
This comment is no longer needed now that the iteration on the union-of-interval (correctly) do not expose the bounds after #1108

Halbaroth commented:
I started a review of the code? Should I discard it?

bclement-ocp commented:
Everything except for changes to the `Intervals.ml` module is still valid and can be reviewed. There are subtleties in the interval module (I am starting to believe that there are pre-existing hard-to-trigger soundness bugs lurking, although the module is very unclear about its invariants so maybe it is actually impossible), so I would hold off reviewing changes to this file specifically. But the rest should not be affected — again, it is just implementation details of the way the Intervals module stores explanations.

Halbaroth commented:
> Everything except for changes to the `Intervals.ml` module is still valid and can be reviewed. There are subtleties in the interval module (I am starting to believe that there are pre-existing hard-to-trigger soundness bugs lurking, although the module is very unclear about its invariants so maybe it is actually impossible), so I would hold off reviewing changes to this file specifically. But the rest should not be affected — again, it is just implementation details of the way the Intervals module stores explanations.

Honestly, when I read this module months ago, my conclusion was we should rewrite it completely. It's not a priority but I plan to work on it later.

bclement-ocp commented:
> Honestly, when I read this module months ago, my conclusion was we should rewrite it completely. It's not a priority but I plan to work on it later.

I am about 99% sure there currently are soundness bugs that can be triggered by properly crafted formulas so I am actually looking into it now. Hopefully I will not need a full rewrite to fix them :/

bclement-ocp committed:
feat(BV): Interval domains for bit-vectors

This patch adds interval domains to the Bitv_rel module, as well as
interreductions between the bitlist and interval domains following:

  Sharpening Constraint Programming approaches for Bit-Vector Theory.
  Zakaria Chihani, Bruno Marre, François Bobot, Sébastien Bardin.
  CPAIOR 2017. International Conference on AI and OR Techniques in
    Constraint Programming for Combinatorial Optimization Problems, Jun
    2017, Padova, Italy.

More precisely:

 - The `Intervals` module is extended to support the `extract`
   operation, which is used to propagate between bit-vector compositions
   and their components;

 - The interreductions are implemented using the new
   `Bitlist.increase_lower_bound`, `Bitlist.decrease_upper_bound`, and
   the new `shared_msb` helper in `Bitv_rel`;

 - Propagations are performed by alternating propagations until fixpoint
   in each domain, followed by interreductions and propagations until
   fixpoint in the other domain, until reaching a fixpoint for the whole
   procedure. It is not clear that this is the best strategy; the goal
   is to try and limit interreductions since they are relatively
   expensive but we should revisit this once more operations are
   supported.

For now, only the `bvule`, `bvult`, `bvugt` and `bvuge` primitives are
supported as built-in bit-vector operations; other operations such as
`bvadd` are still encoded using `bv2nat`. These operations will be
migrated to bit-vector primitives in a follow-up PR.

Finally, there are some tests for the tricky bits (`Intervals.extract`
and the interreduction primitives) using QCheck.

bclement-ocp committed:
Review comments

bclement-ocp commented:
I have adapted the code on top of #1108  (unfortunately I messed up the rebase so the history is a bit weird, apologies for that!).

bclement-ocp commented:
> I am about 99% sure there currently are soundness bugs that can be triggered by properly crafted formulas so I am actually looking into it now. Hopefully I will not need a full rewrite to fix them :/

FWIW: We were in the 1% of chances, but the code in this PR for `Intervals.extract` was still wrong. Fairly confident the new one is right (and if wrong that we have a chance to catch it by reading the code).

Halbaroth commented:
As soon as you rebase it, I will do a last pass on this PR but I think it's already good.

bclement-ocp commented:
Rebased the whole stack.

Halbaroth commented on src/lib/reasoners/rel_utils.ml:
```diff
@@ -267,7 +267,7 @@ module type Domain = sig
+  val map_leaves : (X.r -> t) -> X.r -> t
   (** [map_leaves f r acc] is the "inverse" of [fold_leaves] in the sense that
       it rebuilds a domain for [r] by using [f] to access the domain for each
       of [r]'s leaves. *)
```
This documentation is outdated.

bclement-ocp replied:
Done (took me a minute to realize you meant the `acc` argument no longer exists).

Halbaroth commented on src/lib/reasoners/intervals.mli:
```diff
@@ -30,6 +30,8 @@
 
 open Intervals_intf
 
+val map_bound : ('a -> 'b) -> 'a bound -> 'b bound
```
Please add a documentation :)

bclement-ocp replied:
Done.

Halbaroth commented on src/lib/reasoners/intervals_intf.ml:
```diff
@@ -653,6 +653,10 @@ module type Ring = sig
 
+  val scale : value -> t -> t
+  (** [scale v u] evaluates to {m \{ v \times x \mid x \in S \}} when [u]
+      evaluates to {m S}. *)
```
This function raises an assertion if the scalar is zero. We should mention it in its documentation.

bclement-ocp replied:
Done.

Halbaroth commented on src/lib/reasoners/intervals.mli:
```diff
@@ -43,11 +45,36 @@ module Real : AlgebraicField
+  (** [extract s i j] returns the bits of [s] from position [i] to [j],
+      inclusive.
+
+      Represents the function [fun x -> floor(x / 2^i) % 2^(j - i + 1)].
```
As this function is clear not semialgebraic, we cannot expect that the image of union of intervals is again an union of intervals. We should mention that this map compute an estimation of this image by an union of intervals.

Do you know if we compute the best approximation, that is the smallest union that contains the image?

bclement-ocp replied:
I am not sure I understand your point about semialgebraicity. The image of an (integer) interval by this function is an union of (integer) intervals (in fact either a single interval or an union of two disjoint intervals), so the image of an union of intervals is also an union of intervals.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+        Constraint.propagate_interval
+        Interval_domains.Ephemeral.structural_propagation
+        (QC.pop queue) dom
+    done
```
These function are almost the same. We can factorize them as follows:
```ocaml 
let propagate (type a) (module Eph : Rel_utils.Ephemeral with type t = a)
  propagator queue touched bcs (dom : a) =
  let touch_c c = QC.push queue (Constraint c) in
  let touch r =
    HX.replace touched r ();
    QC.push queue (Structural r);
    Constraints.iter_parents touch_c r bcs
  in
  try
    while true do
      Eph.iter_changed touch dom;
      Eph.clear_changed dom;
      Any_constraint.propagate
        propagator Eph.structural_propagation
        (QC.pop queue) dom
    done
  with QC.Empty -> ()

let propagate_bitlist =
  propagate (module Domains.Ephemeral) Constraint.propagate

let propagate_intervals =
  propagate (module Interval_domains.Ephemeral) Constraint.propagate_interval
``` 
(Assuming that we have extract the signature of the `Ephemeral` in the module type `Rel_utils.Domains`).

bclement-ocp replied:
I simplify this in the patch that I mentioned at the dev meeting (deduplicating domains); I'd prefer merging the these functions in that patch (also to avoid painful rebase) if that's OK.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -149,11 +236,17 @@ module Constraint : sig
   (** [propagate ~ex t dom] propagates the constraint [t] in domain [dom].
 
       The explanation [ex] justifies that the constraint [t] applies, and must
       be added to any domain that gets updated during propagation. *)
```
I think we should rename this function into `propagate_bitlist`

bclement-ocp replied:
Agreed — actually did this in a later patch in the series.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -75,6 +75,93 @@ let is_bv_ty = function
 
 let is_bv_r r = is_bv_ty @@ X.type_info r
 
+module Interval_domain : Rel_utils.Domain with type t = Intervals.Int.t = struct
```
For sake of uniformity, we should rename `Domains` into `Bitlist_domains`.

bclement-ocp replied:
I wanted this patch to avoid touching the bitlist part as much as possible. Renamed it.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+
+    (* First, we propagate the pending constraints to both domains. Changes in
+       the bitlist domain are used to shrink the interval domains. *)
+    touch_pending queue;
+    propagate_bitlist queue touched bcs bdom;
+    assert (QC.is_empty queue);
```
It would be cleaner to move the creation of the queue in `propagate_bitlist`. The assertion can be remove after this modification. So the `propagate` function looks like this:
```ocaml 
let propagate (type a) (module Eph : Rel_utils.Ephemeral with type t = a) propagator =
  let queue = QC.create 17 in
  let touch_c c = QC.push queue (Constraint c)
  in fun touched bcs (dom : a) ->
  QC.clear queue;
  Constraints.iter_pending touch_c bcs;
  let touch r =
    HX.replace touched r ();
    QC.push queue (Structural r);
    Constraints.iter_parents touch_c r bcs
  in
  try
    while true do
      Eph.iter_changed touch dom;
      Eph.clear_changed dom;
      Any_constraint.propagate
        propagator Eph.structural_propagation
        (QC.pop queue) dom
    done
  with QC.Empty -> ()
```
I put the queue in the closure to avoid doing a lot of allocations in the hot loop.

bclement-ocp replied:
> I put the queue in the closure to avoid doing a lot of allocations in the hot loop.

This is the reason I avoided doing something like this — there is enough imperative things and mutable state going on in that part of the code that I really don't want to hide side effects in partial applications. In a situation like this (lots of mutable structures) I think the more repetitive/less elegant code is better because things are laid out more clearly in a sequential way.

Ultimately, the goal is to find the right abstraction and move this code in a shared module that all relations can communicate through. As long as the general domains architecture is still in a bit of flux, I'd really prefer to keep the constraint propagation part as pedestrian as possible.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+    (* Now the interval domain is stable, but the new intervals may have an
+       impact on the bitlist domains, so we must shrink them again when
+       applicable. We repeat until a fixpoint is reached. *)
     let bcs = Constraints.clear_pending bcs in
```
As we flush the pending constraints here, we can use the new `propagate` function in this loop again.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+      ) touched;
+    HX.clear touched;
+    propagate_intervals queue touched bcs idom;
+    assert (QC.is_empty queue);
```
Same here.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+   https://cea.hal.science/cea-01795779/document
+
+   This function is a wrapper calling [Bitlist.increase_lower_bound] and
+   [Bitlist.decrease_upper_bound] on all the constituent interavals of an union;
```
```suggestion
   [Bitlist.decrease_upper_bound] on all the constituent intervals of an union;
```

bclement-ocp replied:
Done.

Halbaroth commented on src/lib/reasoners/bitv_rel.ml:
```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+            Interval_domains.Ephemeral.(!!(handle idom r))
+        ) touched;
+      HX.clear touched;
+      propagate_bitlist queue touched bcs bdom;
```
I don't understand why the queue isn't necessary empty here? We have flushed `bcs` before this loop and `bcs` is immutable and the queue is empty before the loop too.

bclement-ocp replied:
The queue is empty here. If `constrain_bitlist_from_interval` changes the domain of some bitlists, it will be tracked in the `changed` field inside `bdom`, which is used by `propagate_bitlist` to refill the queue.

bclement-ocp committed:
Add _bitlist prefix

bclement-ocp committed:
Fix typo

bclement-ocp committed:
Add documentation

bclement-ocp committed:
Mention exception

bclement-ocp committed:
Remove obsolete argument from doc

bclement-ocp commented:
@Halbaroth as discussed prior to my vacation this is ready for a second pass.

bclement-ocp committed:
Bitlist_domains

Halbaroth reviewed (approved):
Let's merge this one :)
